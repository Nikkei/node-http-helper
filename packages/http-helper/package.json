{
    "name": "@nikkei/http-helper",
    "version": "2.4.1",
    "type": "module",
    "files": [
        "__dist/",
        "LICENSE",
        "package.json",
        "README.md"
    ],
    "typings": "./__dist/mod.d.ts",
    "exports": {
        ".": {
            "types": "./__dist/mod.d.ts",
            "default": "./__dist/mod.js"
        },
        "./auth": {
            "types": "./__dist/auth/mod.d.ts",
            "default": "./__dist/auth/mod.js"
        },
        "./cache": {
            "types": "./__dist/cache/mod.d.ts",
            "default": "./__dist/cache/mod.js"
        },
        "./csp": {
            "types": "./__dist/csp/mod.d.ts",
            "default": "./__dist/csp/mod.js"
        },
        "./sec_fetch": {
            "types": "./__dist/sec_fetch/mod.d.ts",
            "default": "./__dist/sec_fetch/mod.js"
        }
    },
    "scripts": {
        "clean": "npm run clean:dist && npm run clean:tsbuildinfo",
        "clean:dist": "rm -rf __dist/",
        "clean:tsbuildinfo": "rm -rf tsconfig.tsbuildinfo",
        "prebuild": "npm run clean",
        "build": "tsc --project ."
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/Nikkei/node-http-helper.git"
    },
    "author": "Nikkei Inc.",
    "license": "MIT",
    "bugs": {
        "url": "https://github.com/Nikkei/node-http-helper/issues"
    },
    "homepage": "https://github.com/Nikkei/node-http-helper",
    "dependencies": {
        "option-t": "^39.0.2"
    }
}
