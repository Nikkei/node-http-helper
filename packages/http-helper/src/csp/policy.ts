import { isNotNull, type Nullable } from 'option-t/nullable/nullable';

// https://w3c.github.io/webappsec-csp/#grammardef-serialized-policy
export function serializePolicy(...values: Array<Nullable<string>>) {
    const nonNull = values.filter(isNotNull);
    const result = nonNull.join('; ');
    return result;
}
