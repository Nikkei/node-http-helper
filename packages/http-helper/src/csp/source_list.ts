import { type Nullable, isNotNull } from '../internals/nullable.js';

/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-serialized-source-list
 */
export function serializeSourceList(items: Iterable<Nullable<string>>): string {
    const itemsArray = Array.isArray(items) ? items : Array.from(items);
    const nonNull = itemsArray.filter(isNotNull);
    const value = nonNull.join(' ');
    return value;
}
