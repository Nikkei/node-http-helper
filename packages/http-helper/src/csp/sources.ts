export const SOURCE_NONE = `'none'`;

/**
 *  - https://w3c.github.io/webappsec-csp/#grammardef-serialized-source-list
 *  - https://w3c.github.io/webappsec-csp/#grammardef-source-expression
 */
export type SourceExpression =
    | SchemeSource
    | HostSource
    | KeywordSource
    | NonceSource
    | HashSource
    | typeof SOURCE_NONE;

/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-scheme-source
 */
export type SchemeSource = `${string}:`;
export function schemeSource(scheme: string): SchemeSource {
    if (scheme.endsWith(':')) {
        throw new Error(`scheme should not be ends with :, but the actual is ${scheme}`);
    }

    const value: SchemeSource = `${scheme}:`;
    return value;
}

declare const HostSourceMarker: unique symbol;
/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-host-source
 */
export type HostSource = string & { readonly [HostSourceMarker]: unknown };
export function hostSource(host: string): HostSource {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- Cheat to do nominal typing.
    return host as any;
}

export const SOURCE_SELF = "'self'";
export const SOURCE_UNSAFE_INLINE = "'unsafe-inline'";
export const SOURCE_UNSAFE_EVAL = "'unsafe-eval'";
export const SOURCE_STRICT_DYNAMIC = "'strict-dynamic'";
export const SOURCE_UNSAFE_HASHES = "'unsafe-hashes'";
export const SOURCE_REPORT_SAMPLE = "'report-sample'";
export const SOURCE_ALLOW_REDIRECTS = "'unsafe-allow-redirects'";
export const SOURCE_WASM_UNSAFE_EVAL = "'wasm-unsafe-eval'";

/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-keyword-source
 */
export type KeywordSource =
    | typeof SOURCE_SELF
    | typeof SOURCE_UNSAFE_INLINE
    | typeof SOURCE_UNSAFE_EVAL
    | typeof SOURCE_STRICT_DYNAMIC
    | typeof SOURCE_UNSAFE_HASHES
    | typeof SOURCE_REPORT_SAMPLE
    | typeof SOURCE_ALLOW_REDIRECTS
    | typeof SOURCE_WASM_UNSAFE_EVAL;

/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-nonce-source
 */
export type NonceSource = `'nonce-${string}'`;
export function nonceSource(base64Value: string): NonceSource {
    const value: NonceSource = `'nonce-${base64Value}'`;
    return value;
}

type HashSourceAlgorithm = 'sha256' | 'sha384' | 'sha512';
/**
 *  https://w3c.github.io/webappsec-csp/#grammardef-hash-source
 */
export type HashSource = `'${HashSourceAlgorithm}-${string}'`;
export function hashSource(algorithm: HashSourceAlgorithm, base64Value: string): HashSource {
    const value: HashSource = `'${algorithm}-${base64Value}'`;
    return value;
}

declare const URIReferenceMarker: unique symbol;
/**
 *  See https://datatracker.ietf.org/doc/html/rfc3986#section-4.1
 */
export type URIReference = string & { readonly [URIReferenceMarker]: unknown };

export function uriReference(value: string): URIReference {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- Cheat to do nominal typing.
    return value as any;
}
