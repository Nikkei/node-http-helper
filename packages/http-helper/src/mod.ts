export * as Auth from './auth/mod.js';
export * as Cache from './cache/mod.js';
export * as ClearSiteData from './clear_site_data.js';
export * as CSP from './csp/mod.js';
export * as HttpHeaderName from './http_header_name.js';
export * as HttpMethod from './http_method.js';
export * as HttpStatus from './http_status_code.js';
export * as Mime from './mime.js';
export { SecFetchDest, SecFetchMode, SecFetchSite, SecFetchUser } from './sec_fetch/mod.js';
