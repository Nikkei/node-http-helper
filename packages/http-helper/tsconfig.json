{
    "extends": "./tsconfig.base.json",
    "compilerOptions": {
        "outDir": "./__obj"
    },
    "include": [
        // @prettier-ignore
        "src/**/*",
        "__tests__/**/*"
    ]
}
