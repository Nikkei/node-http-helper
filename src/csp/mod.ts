export { serializePolicy } from './policy.js';

export {
    type BaseUriDirective,
    DIRECTIVE_BASE_URI,
    type SandboxDirective,
    DIRECTIVE_SANDBOX,
} from './directives/document.js';
export {
    type ChildSrcDirective,
    DIRECTIVE_CHILD_SRC,
    type ConnectSrcDirective,
    DIRECTIVE_CONNECT_SRC,
    type DefaultSrcDirective,
    DIRECTIVE_DEFAULT_SRC,
    type FontSrcDirective,
    DIRECTIVE_FONT_SRC,
    type FrameSrcDirective,
    DIRECTIVE_FRAME_SRC,
    type ImgSrcDirective,
    DIRECTIVE_IMG_SRC,
    type ManifestSrcDirective,
    DIRECTIVE_MANIFEST_SRC,
    type MediaSrcDirective,
    DIRECTIVE_MEDIA_SRC,
    type ObjectSrcDirective,
    DIRECTIVE_OBJECT_SRC,
    type PrefetchSrcDirective,
    DIRECTIVE_PREFETCH_SRC,
    type ScriptSrcDirective,
    DIRECTIVE_SCRIPT_SRC,
    type ScriptSrcAttrDirective,
    DIRECTIVE_SCRIPT_SRC_ATTR,
    type ScriptSrcElemDirective,
    DIRECTIVE_SCRIPT_SRC_ELEM,
    type StyleSrcDirective,
    DIRECTIVE_STYLE_SRC,
    type StyleSrcAttrDirective,
    DIRECTIVE_STYLE_SRC_ATTR,
    type StyleSrcElemDirective,
    DIRECTIVE_STYLE_SRC_ELEM,
} from './directives/fetch.js';
export {
    type FormActionDirective,
    DIRECTIVE_FORM_ACTION,
    type FrameAncestorsDirective,
    DIRECTIVE_FRAME_ANCESTORS,
} from './directives/navigation.js';
export { type UpgradeInsecureRequestsDirective, DIRECTIVE_UPGRADE_INSECURE_REQUESTS } from './directives/other_spec.js';
export { type WorkerSrcDirective, DIRECTIVE_WORKER_SRC } from './directives/other.js';
export {
    type ReportURIDirective,
    DIRECTIVE_REPORT_URI,
    type ReportToDirective,
    DIRECTIVE_REPORT_TO,
} from './directives/reporting.js';

export { type SerializedDirective, serializeDirective } from './directives/serialize.js';
export { serializeSourceList } from './source_list.js';
export {
    SOURCE_NONE,
    type SourceExpression,
    type SchemeSource,
    schemeSource,
    type HostSource,
    hostSource,
    type KeywordSource,
    SOURCE_SELF,
    SOURCE_UNSAFE_INLINE,
    SOURCE_UNSAFE_EVAL,
    SOURCE_STRICT_DYNAMIC,
    SOURCE_UNSAFE_HASHES,
    SOURCE_REPORT_SAMPLE,
    SOURCE_ALLOW_REDIRECTS,
    SOURCE_WASM_UNSAFE_EVAL,
    type NonceSource,
    nonceSource,
    type HashSource,
    hashSource,
    type URIReference,
    uriReference,
} from './sources.js';
