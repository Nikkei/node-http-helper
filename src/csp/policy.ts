// https://w3c.github.io/webappsec-csp/#grammardef-serialized-policy
export function serializePolicy(...values: Array<string>) {
    const result = values.join('; ');
    return result;
}
