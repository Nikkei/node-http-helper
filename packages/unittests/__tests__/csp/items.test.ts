import { test, expect } from 'vitest';

import * as CSP from '@nikkei/http-helper/csp';

test('CSP exported items', () => {
    expect(CSP).toMatchSnapshot();
});
