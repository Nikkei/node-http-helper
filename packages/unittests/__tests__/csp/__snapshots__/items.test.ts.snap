// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSP exported items 1`] = `
{
  "DIRECTIVE_BASE_URI": "base-uri",
  "DIRECTIVE_CHILD_SRC": "child-src",
  "DIRECTIVE_CONNECT_SRC": "connect-src",
  "DIRECTIVE_DEFAULT_SRC": "default-src",
  "DIRECTIVE_FONT_SRC": "font-src",
  "DIRECTIVE_FORM_ACTION": "form-action",
  "DIRECTIVE_FRAME_ANCESTORS": "frame-ancestors",
  "DIRECTIVE_FRAME_SRC": "frame-src",
  "DIRECTIVE_IMG_SRC": "img-src",
  "DIRECTIVE_MANIFEST_SRC": "manifest-src",
  "DIRECTIVE_MEDIA_SRC": "media-src",
  "DIRECTIVE_OBJECT_SRC": "object-src",
  "DIRECTIVE_PREFETCH_SRC": "prefetch-src",
  "DIRECTIVE_REPORT_TO": "report-to",
  "DIRECTIVE_REPORT_URI": "report-uri",
  "DIRECTIVE_SANDBOX": "sandbox",
  "DIRECTIVE_SCRIPT_SRC": "script-src",
  "DIRECTIVE_SCRIPT_SRC_ATTR": "script-src-attr",
  "DIRECTIVE_SCRIPT_SRC_ELEM": "script-src-elem",
  "DIRECTIVE_STYLE_SRC": "style-src",
  "DIRECTIVE_STYLE_SRC_ATTR": "style-src-attr",
  "DIRECTIVE_STYLE_SRC_ELEM": "style-src-elem",
  "DIRECTIVE_UPGRADE_INSECURE_REQUESTS": "upgrade-insecure-requests",
  "DIRECTIVE_WORKER_SRC": "worker-src",
  "SOURCE_ALLOW_REDIRECTS": "'unsafe-allow-redirects'",
  "SOURCE_NONE": "'none'",
  "SOURCE_REPORT_SAMPLE": "'report-sample'",
  "SOURCE_SELF": "'self'",
  "SOURCE_STRICT_DYNAMIC": "'strict-dynamic'",
  "SOURCE_UNSAFE_EVAL": "'unsafe-eval'",
  "SOURCE_UNSAFE_HASHES": "'unsafe-hashes'",
  "SOURCE_UNSAFE_INLINE": "'unsafe-inline'",
  "SOURCE_WASM_UNSAFE_EVAL": "'wasm-unsafe-eval'",
  "hashSource": [Function],
  "hostSource": [Function],
  "nonceSource": [Function],
  "schemeSource": [Function],
  "serializeDirective": [Function],
  "serializePolicy": [Function],
  "serializeSourceList": [Function],
  "uriReference": [Function],
}
`;
